### ISSUE TITLE
Reduce quality of images to reduce filesize of output (Graphviz installed on server)

### ISSUE THREAD

**Issue by Neriderc** [Neriderc/GVExport#395]
In the context of future to-dos concerning photos please consider the problem, that PDFs including photos often get much bigger than diagram-PDFs without photos. I observed cases for the same diagram with photo 10MB but only 100KB without. You can compress such a PDF without noticable loss in quality by a PDF editor down below 1MB.
I know that the reason for big PDF-diagrams are too big jpg-files and I remember, that you already explained that it is impossible to get thumbnail-jpg directly from webtrees.

_Originally posted by @schuco in https://github.com/Neriderc/GVExport/issues/393#issuecomment-1474132082_

**Comment by Neriderc** [comment]
@schuco responding to your comment in #393:
> But if you have a mixture of photos with very different size and basic resolution a general reduction of DPI may not be good in all cases.

Don't think of the DPI setting as being for the images, it is for the diagram. How images are treated depends on the exact settings, but for PDF output where Graphviz is installed on the server, photos are scaled to be the width of the DPI setting (e.g. if DPI is the default of 72, each photo will be scaled to 72 pixels wide). 

I have created this as a separate issue as I think this is separate, but also I think we have a clearer path here. There is a PHP function imagejpeg(...) that allows to set the JPG quality. We can make all images JPG and drop the quality setting, which is something I've only just found we can do somewhat easily (I think).

This does raise a question of if this should be yet another setting or if there is a sensible default here. We could set this to (say) 65 (default is 75), then if DPI is increased about the default, we could also add to this in order to allow users to get higher quality images if they want. 

Personally I'm a fan of adding new settings for everything, letting users customise to their heart's content (knowing that these are all hidden behind the advanced settings sections). But I know @hartenthaler is concerned about adding too many settings, which is also valid. I think this could be handled by a bit of a rethink of how settings are displayed, so I've added #396 to discuss this.

In terms of this quality setting, I think we should add it as an option. We could allow the direct setting of it (0-100 with 75 being the PHP default value), or we could use a dropdown with "Very high", "High", "Normal", "Low", and "Very low" in order to make it easier for users.

**Comment by Neriderc** [comment]
I have now implemented this by adding two new options that are only available when you have Graphviz installed on the server, and when you have selected either "SVG" or "PDF" options. Other file types don't embed images, so we don't have the same control (or the same need).

If you don't have Graphviz installed, there isn't a clear path on how to implement this for SVG as the images are loaded from webtrees rather than from the hard drive. PDF output without Graphviz on the server is not a true PDF conversion, it is just a JPEG image of the whole diagram embedded in a PDF document, so this isn't really the same. The DPI option should be used in this case.

The JPEG quality settings run between 0 and 100. I have made a dropdown list with named values, with Medium being the default (set to 50, which is lower than the JPG default of 75 that would have been used previously - the setting High is equal to the 75 setting).

I have also added an option to convert images to JPG. PHP seems to support PNG, BMP, and GIF in additional to JPG, so these can all be converted to JPG and apply the quality setting. Other files will be embedded as is. By default, converting all images to JPG is disabled.

Here's how it looks. I've also merged this into the main branch so feel free to try it out and let me know what you think.
![image](https://user-images.githubusercontent.com/89974636/235103706-4289559e-83f7-41ca-abf1-b14cb412f564.png)

**Comment by schuco** [comment]
I compared different quality levels for a quite large diagram with many photos
highest     4.3 MB
medium    3.8 MB
low            3.0 MB
lowest       566 KB
lowest results in really poor pictures, low seems to be sufficient at a first glance.
Thereafter I used the compression tool in my pdf editor (PDF Architect) which compressed from 3.0 MB to 303 KB, still with sufficient quality.
So the new approach in GVE will help to download a PDF, but can still be compressed by other tools quite a lot.

**Comment by Neriderc** [comment]
Perhaps PDF Architect has access to functions that we don't (or don't know about), but maybe the balance isn't right. Probably PDF Architect staff have played around with combinations of resolution and quality to get a good balance.

I tried changing the DPI setting. On my test diagram at 72 DPI and Medium photo quality (50), the file was 2.4 MB. I changed the DPI to 30 and the file changed to 559 KB, less than 1/4 of the size. From my testing, it seems the quality setting has a much bigger impact at higher DPIs than lower ones, at a DPI of 30 the file size barely changed between the different settings.

The DPI setting doesn't impact the text/diagram components in the native PDF as these components are drawn as vector images, so you can zoom in, and they will always look crisp. This is not the case if you for example export the diagram as a JPG or PNG. In these cases, the DPI needs to be high enough that you can read the text. In my view, this means we shouldn't change the default DPI to a lower value.

So I think there are some options here.

1. Reduce the DPI to (say) 30 when a vector format is chosen (I don't like this idea, we are changing a setting that the user may have already set)
2. Change photos to be 1/2 the DPI wide instead of the same width as the DPI
3. Add a new option to adjust the resolution of images. We could use width/height, but we don't have this control when _displaying_ the image (i.e. making it bigger/smaller), only using this for the resolution. This could get confusing.
4. Add new option as per 3, but do it as a % so it's not as confusing.
5. Do nothing, users can use DPI setting to adjust photo resolution when using SVG or server side PDF

I think my preference is to do 2 and also 4. Halve the photo resolution but let the user adjust this with a %.

Thoughts?

**Comment by schuco** [comment]
Maybe the strategy of PDF-Architect's compression is similar to your option 2. The attached screenshot from the compression I talked about before 
> I compared different quality levels for a quite large diagram with many photos highest 4.3 MB medium 3.8 MB low 3.0 MB lowest 566 KB lowest results in really poor pictures

shows that  compression to medium file size aims on picture resampling to 600 dpi:
![grafik](https://user-images.githubusercontent.com/5896925/235300014-323aa676-a0d2-4d41-a1f7-63fa00ef7aee.png)

I don't know anything about internal precedures in the GVE module, but it seems to me that you manipulate photos between taking them from the webtrees db and providing them for the graphic. 
Would it be possible to manipulate only those photos above a certain size, e.g greater than 300KB, and compress them to be max. 300 KB?
There should be an internal procedure applied in webtrees to generate thumbnails. The size of thumbnails would be sufficient for the diagrams.

**Comment by schuco** [comment]
Today I made a diagram including very large photos (family at weddings). The downloaded pdf (graphviz serverside) was about 14 MB which was compressed to 11 MB by PDF-Architect only.
In this case using the new GVE feature with lowest JPG quality resulted in only about 300 KB for the downloaded file.
Clearly the new feature has an advantage here!

**Comment by Neriderc** [comment]
> I don't know anything about internal precedures in the GVE module, but it seems to me that you manipulate photos between taking them from the webtrees db and providing them for the graphic. Would it be possible to manipulate only those photos above a certain size, e.g greater than 300KB, and compress them to be max. 300 KB? There should be an internal procedure applied in webtrees to generate thumbnails. The size of thumbnails would be sufficient for the diagrams.

The issue here is that when Graphviz is installed on the server, the photos do not come from webtrees. Instead, we ask webtrees for the location of the photo on the hard drive and then grab the file directly. This is because Graphviz needs the direct location of the file, there isn't a way to tell Graphviz to ask webtrees for it.

And if each photo was 300 KB, then 100 photos would make the file 30000 KB or 30 MB. Instead, we are scaling the resolution to be the same for each, and then setting the quality setting, which is making each photo much smaller than 300 KB I presume, from testing a diagram with about 50 photos at 72DPI, it's about 5.5 MB so about 100 KB per photo. This is with quality set to High (75 - the previous default).

> Today I made a diagram including very large photos (family at weddings). The downloaded pdf (graphviz serverside) was about 14 MB which was compressed to 11 MB by PDF-Architect only. In this case using the new GVE feature with lowest JPG quality resulted in only about 300 KB for the downloaded file. Clearly the new feature has an advantage here!

Wow! Was this also with a low DPI setting or the default 72?

**Comment by schuco** [comment]
I compared effects of dpi-setting versus jpg compression:
![grafik](https://user-images.githubusercontent.com/5896925/235756541-d2d2aef8-72e4-4019-a02a-c053407dd079.png)
Comparing photo quality subjectively as a function of dpi and jpg-compression it seems more effective to reduce dpi rather than jpg quality:
high quality 40dpi = 2.575kB
![grafik](https://user-images.githubusercontent.com/5896925/235760277-4bd4222b-99be-4e84-bd05-d8c966b4f109.png)
high quality 50dpi = 3.799kB
![grafik](https://user-images.githubusercontent.com/5896925/235760551-7e68f139-f63f-406f-94c7-ed23324592c5.png)
low quality 40 dpi = 2.028kB
![grafik](https://user-images.githubusercontent.com/5896925/235758908-ae326000-8fa6-4b1a-b611-c75b5ff9295a.png)
low quality 50dpi = 2.925kB
![grafik](https://user-images.githubusercontent.com/5896925/235760800-319a95c5-aee1-4e22-9acb-acfe61fa196d.png)

**Comment by Neriderc** [comment]
Sorry, I would have replied earlier but have spent the last week or more dealing with what turned out to be a failed hard drive on my server. One morning it just wouldn't turn on, then when I got a chance I grabbed a spare hard drive and tried restoring my backup, and spent a long time fighting with it. I thought maybe the backup was corrupt, but it turned out the spare hard drive was failing and so was regularly failing to write things to the disk... I ended up buying a new hard drive and after restoring my backup I managed to get everything up and running again!

Anyway, thanks for all that work comparing photos. I think we should do the options 2 and 4 as I suggested above. I will have made a new issue for this, to allow some time for feedback (also discussions on closed issues like this are pretty hidden from anyone not getting email notifications).
